### PR TITLE
v0.8 release

### PR DESCRIPTION
**Pull request by blazekadam**
**At least 2 reviews are required here!**

### CODE REVIEW

File: cxflow/main_loop.py [hunk 1]
```diff
@@ -1,73 +1,84 @@
...
+    """Predict stream name."""
 
     def __init__(self,   # pylint: disable=too-many-arguments
-                 net: AbstractNet, dataset: AbstractDataset,
-                 hooks: typing.Iterable[AbstractHook]=(),
-                 extra_streams: typing.List[str]=(),  # pylint: disable=invalid-sequence-index
+                 model: AbstractModel, dataset: AbstractDataset,
+                 hooks: Iterable[AbstractHook]=(),
+                 extra_streams: List[str]=(),  # pylint: disable=invalid-sequence-index
                  on_unused_sources: str = 'warn',
```
Comment (petrbel) [inline cxflow/main_loop.py]: extra spaces

File: cxflow/main_loop.py [hunk 2]
```diff
@@ -1,73 +1,84 @@
...
-                 fixed_batch_size: int = None,
+                 fixed_batch_size: Optional[int] = None,
```
Comment (petrbel) [inline cxflow/main_loop.py]: extra spaces

File: cxflow/main_loop.py [hunk 3]
```diff
@@ -1,73 +1,84 @@
...
 
 class MainLoop:   # pylint: disable=too-many-instance-attributes
-    """Train the network, manage hooks etc."""
+    """**cxflow** main loop for training and model inference."""
 
-    UNUSED_SOURCE_ACTIONS = {'ignore', 'warn', 'error'}
+    UNUSED_SOURCE_ACTIONS = ['ignore', 'warn', 'error']
+    """Possible actions to be taken when a stream source is unused by the trained model."""
+
+    TRAIN_STREAM = 'train'
```
Comment (petrbel) [inline cxflow/main_loop.py]: maybe this could be also configurable by `main_loop` config section?

File: cxflow/main_loop.py [hunk 4]
```diff
@@ -1,73 +1,84 @@
...
-        self._all_streams = list(self._extra_streams) + ['train']
+        self._extra_streams = list(extra_streams)
         self._skip_zeroth_epoch = skip_zeroth_epoch
 
-    def _create_epoch_data(self):
+    def _create_epoch_data(self) -> AbstractHook.EpochData:
         """Create empty epoch data double dict."""
-        return OrderedDict([(stream_name, OrderedDict()) for stream_name in self._all_streams])
+        return OrderedDict([(stream_name, OrderedDict())
+                            for stream_name in self._extra_streams + [MainLoop.TRAIN_STREAM]])
```
Comment (petrbel) [inline cxflow/main_loop.py]: swap the order (train shold be first)

File: cxflow/main_loop.py [hunk 5]
```diff
@@ -100,65 +114,110 @@ def _run_epoch(self, stream: AbstractDataset.Stream, train: bool, stream_name: s
...
+        for hook in self._hooks:
+            hook.before_training()
 
         try:
-            epoch_id = 0
+            run_func()
+        except TrainingTerminated as ex:
+            logging.info('Training terminated by a hook: %s', ex)
+        except KeyboardInterrupt:
+            logging.warning('Training terminated by a keyboard interrupt')
```
Comment (petrbel) [inline cxflow/main_loop.py]: also predict might be interrupted...

File: cxflow/models/abstract_model.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+from abc import abstractmethod, ABCMeta
+from typing import Iterable, Mapping, Optional
+
+from ..datasets import AbstractDataset
+
+
+class AbstractModel(metaclass=ABCMeta):
+    """
+    Abstract machine learning model which exposes input and output names, run and save methods.
+    `AbstractModel` implementations are trainable with ``cxflow.MainLoop``.
```
Comment (petrbel) [inline cxflow/models/abstract_model.py]: double backticks

File: cxflow/models/abstract_model.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        :param name_suffix: name suffix to be appended to the saved model.
+        :return: path to the saved file/dir
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def restore_fallback(self) -> str:
+        """
+        Return the fully-qualified name of the fallback restore class (e.g. `module.submodule.BaseClass`).
```
Comment (petrbel) [inline cxflow/models/abstract_model.py]: double backticks

File: cxflow/models/abstract_model.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+        When restoring a model, cxflow tries to use the fallback class if the construction of the model object specified
+        in `model` configuration section fails.
```
Comment (petrbel) [inline cxflow/models/abstract_model.py]: double backticks

File: cxflow/tests/cli/common_test.py [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+
+
+class DummyDataset:
+    """Dummy dataset which loads the given config to self.config."""
+    def __init__(self, config_str):
+        self.config = yaml.load(config_str)
+
+
+class DummyHook(AbstractHook):
+    """Dummy hook which save its **kwargs to self.kwargs."""
```
Comment (petrbel) [inline cxflow/tests/cli/common_test.py]: doble backticks (2x)

File: cxflow/tests/cli/common_test.py [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+
+class DummyModelWithKwargs2(DummyModelWithKwargs):
+    """Direct inheritor of `DummyModelWithKwargs`.
+
+    For restoring purposes only."""
+    pass
+
+
+class CLICommonTest(CXTestCaseWithDir):
+    """cli common test case."""
```
Comment (petrbel) [inline cxflow/tests/cli/common_test.py]: cli -> CLI

File: cxflow/tests/cli/common_test.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+        expected_config = {'batch_size': 10, 'output_dir': 'dummy_dir'}
+
+        dataset = create_dataset(config=config, output_dir='dummy_dir')
+
+        self.assertTrue(isinstance(dataset, DummyDataset))
+        self.assertTrue(hasattr(dataset, 'config'))
+        self.assertDictEqual(dataset.config, expected_config)
+
+    def test_create_hooks(self):
+        """Test hooks creation in train_create_hooks."""
```
Comment (petrbel) [inline cxflow/tests/cli/common_test.py]: double backticks

File: cxflow/tests/hooks/catch_sigint_hook_test.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+import os
+import signal
+
+from cxflow.tests.test_core import CXTestCase
+from cxflow.hooks.catch_sigint import CatchSigint
+from cxflow.hooks.abstract_hook import TrainingTerminated
+
+
+class SigintHookTest(CXTestCase):
+    """Test case for CatchSigint hook."""
```
Comment (petrbel) [inline cxflow/tests/hooks/catch_sigint_hook_test.py]: double backticks

File: cxflow/tests/hooks/catch_sigint_hook_test.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    def _sigint_handler(self, *_):
+        self._sigint_unhandled = True
+
+    def setUp(self):
+        """Register the _sigint_handler."""
+        signal.signal(signal.SIGINT, self._sigint_handler)
+        self._sigint_unhandled = False
+
+    def test_before_training(self):
+        """Test SigintHook does not handle the sigint before the training itself."""
```
Comment (petrbel) [inline cxflow/tests/hooks/catch_sigint_hook_test.py]: double backticks

File: cxflow/tests/hooks/catch_sigint_hook_test.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+        CatchSigint()
+        os.kill(os.getpid(), signal.SIGINT)
+        self.assertTrue(self._sigint_unhandled, 'SigintHook handles SIGINT before training')
+
+    def test_inside_training_no_raise(self):
+        """Test SigintHook does handle the sigint during the training."""
```
Comment (petrbel) [inline cxflow/tests/hooks/catch_sigint_hook_test.py]: double backticks

File: cxflow/tests/hooks/catch_sigint_hook_test.py [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+        hook = CatchSigint()
+        hook.before_training()
+        os.kill(os.getpid(), signal.SIGINT)
+        self.assertFalse(self._sigint_unhandled, 'SigintHook does not handle SIGINT while training')
+        # double SIGINT cannot be easily tested (it quits(1))
+
+    def test_inside_training_raise(self):
+        """Test SigintHook does rise TrainingTerminated exception."""
```
Comment (petrbel) [inline cxflow/tests/hooks/catch_sigint_hook_test.py]: double backticks

File: cxflow/tests/hooks/catch_sigint_hook_test.py [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+        with self.assertRaises(TrainingTerminated):
+            hook = CatchSigint()
+            hook.before_training()
+            os.kill(os.getpid(), signal.SIGINT)
+            hook.after_batch()
+
+    def test_after_training(self):
+        """Test SigintHook does not handle the sigint after the training."""
```
Comment (petrbel) [inline cxflow/tests/hooks/catch_sigint_hook_test.py]: double backticks

File: cxflow/tests/hooks/check_hook_test.py [hunk 17]
```diff
@@ -1,12 +1,12 @@
 """
-Test module for train check hook (cxflow.hooks.train_check_hook).
+Test module for train check hook (cxflow.hooks.check_hook).
```
Comment (petrbel) [inline cxflow/tests/hooks/check_hook_test.py]: double backticks

File: cxflow/tests/hooks/compute_stats_hook.py [hunk 18]
```diff
@@ -1,16 +1,16 @@
...
 import collections
 
 from cxflow.tests.test_core import CXTestCase
-from cxflow.hooks.stats_hook import StatsHook
+from cxflow.hooks.compute_stats import ComputeStats
 
 
 class StatsHookTest(CXTestCase):
-    """Test case for StatsHook."""
+    """Test case for ComputeStats hook."""
```
Comment (petrbel) [inline cxflow/tests/hooks/compute_stats_hook.py]: double backticks

File: cxflow/tests/hooks/log_profile_hook_test.py [hunk 19]
```diff
@@ -1,46 +1,46 @@
...
+                                 'read_batch_test': [1.1, 1.11, 1.111],
                                  'after_batch_hooks_test': [2.2, 2.22, 2.222],
                                  'eval_batch_test': [3.3, 3.33, 3.333],
                                  'after_epoch_hooks': [5.3]}
 
 
 class ProfileHookTest(CXTestCase):
     """
-    Test case for ProfileHook.
+    Test case for LogProfile hook.
```
Comment (petrbel) [inline cxflow/tests/hooks/log_profile_hook_test.py]: double backticks

File: cxflow/tests/hooks/log_profile_hook_test.py [hunk 20]
```diff
@@ -1,46 +1,46 @@
...
 
     Hereby, we test only proper handling (logging) of the profile.
     The profiling itself is tested in the main_loop test (see cxflow.tests.main_loop_test.py).
```
Comment (petrbel) [inline cxflow/tests/hooks/log_profile_hook_test.py]: double backticks

File: cxflow/tests/hooks/save_hook_test.py [hunk 21]
```diff
@@ -0,0 +1,146 @@
+"""
+Test module for saver hook (cxflow.hooks.save_hook).
```
Comment (petrbel) [inline cxflow/tests/hooks/save_hook_test.py]: double backticks

File: cxflow/tests/hooks/save_hook_test.py [hunk 22]
```diff
@@ -0,0 +1,146 @@
...
+        ('valid', collections.OrderedDict([
+            ('accuracy', 0.8),
+            ('loss', collections.OrderedDict([('mean', valid_loss_mean_val)])),
+        ]))
+    ])
+    return epoch_data
+
+
+class EmptyModel(AbstractModel):
+    """The model which raises `ValueError` if save method is called."""
```
Comment (petrbel) [inline cxflow/tests/hooks/save_hook_test.py]: double backticks

File: cxflow/tests/hooks/save_hook_test.py [hunk 23]
```diff
@@ -0,0 +1,146 @@
...
+    def output_names(self) -> List[str]:
+        pass
+
+    @property
+    def restore_fallback(self) -> str:
+        return ''
+
+
+class SaverHookTest(CXTestCase):
+    """Test case for SaverHook."""
```
Comment (petrbel) [inline cxflow/tests/hooks/save_hook_test.py]: double backticks

File: cxflow/tests/hooks/save_hook_test.py [hunk 24]
```diff
@@ -0,0 +1,146 @@
...
+
+    def test_raise_on_save_failure(self):
+        """
+        Test raising an exception if `on_save_failure`
```
Comment (petrbel) [inline cxflow/tests/hooks/save_hook_test.py]: double backticks

File: cxflow/tests/hooks/save_hook_test.py [hunk 25]
```diff
@@ -0,0 +1,146 @@
...
+        """Test saving/not saving every n epoch."""
+        with self.assertRaises(IOError):
+            SaveEvery(model=EmptyModel(), n_epochs=3,
+                      on_failure='error').after_epoch(epoch_id=30)
+        SaveEvery(model=EmptyModel(), n_epochs=3).after_epoch(epoch_id=29)
+
+    def test_no_raise(self):
+        """
+        Test whether an exception is not raised
+        if `on_save_failure` is set to warn/ignore.
```
Comment (petrbel) [inline cxflow/tests/hooks/save_hook_test.py]: double backticks

File: cxflow/tests/hooks/save_hook_test.py [hunk 26]
```diff
@@ -0,0 +1,146 @@
...
+        SaveEvery(model=EmptyModel(), n_epochs=3,
+                  on_failure='ignore').after_epoch(epoch_id=30)
+
+
+class BestSaverHookTest(CXTestCase):
+    """Test case for SaveBest hook."""
+
+    def test_raise_invalid_on_save_failure(self):
+        """
+        Test raising an exception if `on_save_failure`
```
Comment (petrbel) [inline cxflow/tests/hooks/save_hook_test.py]: double backticks

File: cxflow/tests/hooks/stop_after_hook_test.py [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+from cxflow.main_loop import MainLoop
+from cxflow.tests.test_core import CXTestCase
+from cxflow.hooks.stop_after import StopAfter
+from cxflow.hooks.abstract_hook import TrainingTerminated
+
+NOTRAIN_STREAM_NAME = 'valid'
+assert MainLoop.TRAIN_STREAM is not NOTRAIN_STREAM_NAME
+
+class EpochStopperHookTest(CXTestCase):
+    """Test case for StopAfter hook."""
```
Comment (petrbel) [inline cxflow/tests/hooks/stop_after_hook_test.py]: double backticks

File: cxflow/tests/hooks/write_csv_hook_test.py [hunk 28]
```diff
@@ -1,5 +1,5 @@
 """
-Test module for csv hook (cxflow.hooks.csv_hook).
+Test module for csv hook (cxflow.hooks.write_csv_hook).
```
Comment (petrbel) [inline cxflow/tests/hooks/write_csv_hook_test.py]: double backticks

File: cxflow/tests/hooks/write_csv_hook_test.py [hunk 29]
```diff
@@ -37,34 +37,34 @@ def _get_epoch_data():
 
 
 class CSVHookTest(CXTestCase):
-    """Test case for CSVHook."""
+    """Test case for WriteCSV hook."""
```
Comment (petrbel) [inline cxflow/tests/hooks/write_csv_hook_test.py]: double backticks

File: cxflow/utils/profile.py [hunk 30]
```diff
@@ -11,21 +11,23 @@ class Timer:
...
-                                 typing.Dict[str, typing.List[float]])   # pylint: disable=invalid-sequence-index
+
+    TimeProfile = typing.Dict[str, typing.List[float]]   # pylint: disable=invalid-sequence-index
 
     def __init__(self, name: str, profile: TimeProfile):
         """
         Create new Timer instance.
+
         :param name: event name under which the measured time should be saved
         :param profile: profile = dict of lists of timings
```
Comment (petrbel) [inline cxflow/utils/profile.py]: remove "profile = "

File: cxflow/cli/args.py [hunk 31]
```diff
@@ -0,0 +1,63 @@
...
+    # create resume subparser
+    resume_parser = subparsers.add_parser('resume', description='Resume cxflow training from the ``config_path``.')
+    resume_parser.set_defaults(subcommand='resume')
+    resume_parser.add_argument('config_path', help='path to the config file or the directory in which it is stored')
+    resume_parser.add_argument('restore_from', nargs='?', default=None,
+                               help='information passed to the model constructor (backend-specific); '
+                                    'usually a directory in which the trained model is stored')
+
+    # create predict subparser
+    predict_parser = subparsers.add_parser('predict')
```
Comment (blazekadam) [inline cxflow/cli/args.py]: add description

File: cxflow/cli/args.py [hunk 32]
```diff
@@ -0,0 +1,63 @@
...
+    predict_parser.set_defaults(subcommand='predict')
+    predict_parser.add_argument('config_path', help='path to the config file or the directory in which it is stored')
+    predict_parser.add_argument('restore_from', nargs='?', default=None,
+                                help='information passed to the model constructor (backend-specific); usually a '
+                                     'directory in which the trained model is stored')
+
+    # create dataset subparser
+    dataset_parser = subparsers.add_parser('dataset')
```
Comment (blazekadam) [inline cxflow/cli/args.py]: add description

File: cxflow/cli/args.py [hunk 33]
```diff
@@ -0,0 +1,63 @@
...
+    dataset_parser.set_defaults(subcommand='dataset')
+    dataset_parser.add_argument('method', help='name of the method to be invoked')
+    dataset_parser.add_argument('config_file', help='path to the config file')
+
+    # create grid-search subparser
+    gridsearch_parser = subparsers.add_parser('gridsearch')
```
Comment (blazekadam) [inline cxflow/cli/args.py]: add description

File: cxflow/cli/common.py [hunk 34]
```diff
@@ -0,0 +1,271 @@
...
+        model = create_object(model_module, model_class, kwargs=model_kwargs)
+    except (ImportError, AttributeError) as ex:
+        if restore_from is None:  # training case
+            raise ImportError('Cannot create model from the specified model module `{}` and class `{}`.'.format(
+                model_module, model_class)) from ex
+
+        else:  # restore cases (resume, predict)
+            logging.warning('Cannot create model from the specified model class `%s`.', model_config['class'])
+            assert 'restore_fallback' in model_config, '`model.restore_fallback` not present in the config'
+            logging.info('Trying to restore with fallback `{}` instead.'.format(model_config['restore_fallback']))
```
Comment (blazekadam) [inline cxflow/cli/common.py]: use %s

File: cxflow/cli/dataset.py [hunk 35]
```diff
@@ -0,0 +1,49 @@
...
+    :param config_path: path to the config file or the directory in which it is stored
+    :param method_name: name of the method to be invoked on the specified dataset
+    :param cl_arguments: additional command line arguments which will update the configuration
+    :param output_root: output root in which the training directory will be created
+    """
+
+    config = dataset = method = output_dir = None
+
+    try:
+        assert path.exists(config_path), '`{}` does not exist'.format(config_path)
```
Comment (blazekadam) [inline cxflow/cli/dataset.py]: Use find_config here

File: cxflow/cli/grid_search.py [hunk 36]
```diff
@@ -14,9 +14,9 @@ def _build_grid_search_commands(script: str, params: typing.Iterable[str]) -> ty
     """
     Build all grid search parameter configurations.
 
-    :param script: String of command prefix, e.g. `cxflow train -v -o log`.
-    :param params: Iterable collection of strings in standard cxflow param form, e.g. 'numerical_param:int=[1, 2]' or
-                   'text_param:str=["hello", "cio"]'.
+    :param script: String of command prefix, e.g. ``cxflow train -v -o log``.
+    :param params: Iterable collection of strings in standard cxflow param form, e.g. ``'numerical_param:int=[1, 2]'``
```
Comment (blazekadam) [inline cxflow/cli/grid_search.py]: remove type hints

File: cxflow/cli/grid_search.py [hunk 37]
```diff
@@ -45,10 +45,10 @@ def grid_search(script: str, params: typing.Iterable[str], dry_run: bool=False)
     """
     Build all grid search parameter configurations and optionally run them.
 
-    :param script: String of command prefix, e.g. `cxflow train -v -o log`.
-    :param params: Iterable collection of strings in standard cxflow param form, e.g. 'numerical_param:int=[1, 2]' or
-                   'text_param:str=["hello", "cio"]'.
-    :param dry_run: If set to true, the built commands will only printed instead of executed.
+    :param script: String of command prefix, e.g. ``cxflow train -v -o log``.
+    :param params: Iterable collection of strings in standard cxflow param form, e.g. ``'numerical_param:int=[1, 2]'``
+                   or ``'text_param:str=["hello", "cio"]'``.
```
Comment (blazekadam) [inline cxflow/cli/grid_search.py]: ditto

File: cxflow/datasets/__init__.py [hunk 38]
```diff
@@ -1,4 +1,8 @@
-"""cxflow dataset module."""
+"""Module with **cxflow** dataset concept (:py:class:`AbstractDataset`) and python :py:class:`BaseDataset`."""
 
 from .abstract_dataset import AbstractDataset
 from .base_dataset import BaseDataset
+
+AbstractDataset.__module__ = '.datasets'
```
Comment (blazekadam) [inline cxflow/datasets/__init__.py]: Set up the module for BaseDataset. check the links above.

File: cxflow/datasets/base_dataset.py [hunk 39]
```diff
@@ -1,66 +1,47 @@
...
+class BaseDataset(AbstractDataset, metaclass=ABCMeta):
     """
     Base class for datasets written in python.
 
     In the inherited class, one should:
-        - override the _init_with_kwargs method insead of __init__
-        - override the create_train_stream method
-        - override any additional create_[stream_name]_stream method in order to make [stream_name] stream available
+        - override the ``_init_with_kwargs`` method instead of ``__init__``
+        - override the ``train_stream`` method
```
Comment (blazekadam) [inline cxflow/datasets/base_dataset.py]: Update docstring

File: cxflow/hooks/__init__.py [hunk 40]
```diff
@@ -1,11 +1,29 @@
...
-from .logging_hook import LoggingHook
-from .profile_hook import ProfileHook
-from .saver_hook import SaverHook, BestSaverHook
-from .sigint_hook import SigintHook
-from .stats_hook import StatsHook
-from .train_check_hook import TrainCheckHook
+"""
+Module with official **cxflow** hooks.
+
+
```
Comment (blazekadam) [inline cxflow/hooks/__init__.py]: extra empty line

File: cxflow/hooks/__init__.py [hunk 41]
```diff
@@ -1,11 +1,29 @@
...
+from .accumulate_variables import AccumulateVariables
+from .write_csv import WriteCSV
+from .stop_after import StopAfter
+from .log_variables import LogVariables
+from .log_profile import LogProfile
+from .save import SaveEvery, SaveBest
+from .catch_sigint import CatchSigint
+from .compute_stats import ComputeStats
+from .check import Check
+
```
Comment (blazekadam) [inline cxflow/hooks/__init__.py]: Set up __module__ for at least AbstractHook

File: cxflow/datasets/__init__.py (hunk 38 above)
Comment (petrbel) [reply cxflow/datasets/__init__.py]: done

File: cxflow/hooks/__init__.py (hunk 41 above)
Comment (petrbel) [reply cxflow/hooks/__init__.py]: done

**Pull request merged**